hathor/feature_activation/validation.py

```diff
@@ -0,0 +1,87 @@
+#  Copyright 2023 Hathor Labs
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from collections import defaultdict
+from typing import Dict, List, Optional, Tuple
+
+from hathor.feature_activation.exception import InvalidFeaturesConfigurationException
+from hathor.feature_activation.model.criteria import Criteria
+
+
+def validate_feature_list(features: List[Criteria]) -> None:
+    """Validates a list of features."""
+    _validate_duplicate_names(features)
+    _validate_conflicting_bits(features)
+
+
+def _validate_duplicate_names(features: List[Criteria]) -> None:
+    """Validates that no two features share the same name."""
+    feature_names = [criteria.name for criteria in features]
+
+    if len(feature_names) != len(set(feature_names)):
+        raise InvalidFeaturesConfigurationException('Feature names should be unique')
+
+
+def _validate_conflicting_bits(features: List[Criteria]) -> None:
+    """
+    Validates that a bit is only reused if the start_height of a new feature is
+    greater than the timeout_height of the previous feature that used that bit.
+    """
+    intervals_by_bit = _get_intervals_by_bit(features)
+    intervals_has_overlaps = [
+        _has_any_overlap(intervals)
+        for intervals in intervals_by_bit.values()
+    ]
+
+    if any(intervals_has_overlaps):
+        raise InvalidFeaturesConfigurationException(
+            'One or more Features have the same bit configured for an overlapping interval'
+        )
+
+
+def _get_intervals_by_bit(features: List[Criteria]) -> Dict[int, List[Tuple[int, int]]]:
```

Comment:
Should we create a `IntervalByBit` namedtuple as below? In this case, we would be able to add the feature and improve the error message in case of overlap.

```python
class IntervalByBit(NamedTuple):
    begin: int
    end: int
    feature: Feature
```

Reply:
Done in https://github.com/HathorNetwork/hathor-core/pull/611/commits/b141815b3655383aa17f7df5c33cba323cb58c19

Reply:
Thanks!